### PR TITLE
Fh 4027 metrics

### PR DESCRIPTION
**Pull request by maleck13**
adds a metric gathering background task. Adds a metrics endpoint for services. Adds keycloak and sync metrics

### CODE REVIEW

**Comment by david-martin** [comment]
Added 17 graphs for fh-sync-server in total. These can be joined (multiple lines on a single graph) in future if needed (and where appropriate).

@maleck13 would you mind reviewing the test as well, added in fa32fec?

image
image
image

File: pkg/mobile/metrics/fh-sync-server.go [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+	"github.com/Sirupsen/logrus"
+	"github.com/feedhenry/mcp-standalone/pkg/mobile"
+	"github.com/pkg/errors"
+)
+
+// TODO: Add multi-worker support?
+type statsResponse struct {
+	Metrics struct {
+		CPUUsage struct {
```
Comment (maleck13) [inline pkg/mobile/metrics/fh-sync-server.go]: As these usage structs are the same, they can be abstracted

```
type Usage struct{
 Current string `json:"current"`
}

type CPUUsage struct{
    Master *Usage `json:"master"`
}

```

etc. Can discuss further in person

File: pkg/mobile/metrics/fh-sync-server.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+	if err := decoder.Decode(&stats); err != nil {
+		return &statsResponse{}, errors.Wrap(err, "failed to decode stats response")
+	}
+	ss.logger.Debugf("decoded stats response %v", stats)
+
+	return &stats, nil
+}
+
+func stringToInt64(val string) (int64, error) {
+	// string expected in format "00.00ms"
```
Comment (maleck13) [inline pkg/mobile/metrics/fh-sync-server.go]: do we want to test the format with a regex or maybe test the length of the split string?

File: pkg/mobile/metrics/fh-sync-server.go [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+	stats, err := ss.getStats(host)
+	if err != nil {
+		return nil, errors.Wrap(err, "fh-sync-server gather failed to list existing services")
+	}
+
+	var ssMetrics = []*metric{}
+	if nil != stats {
+		// API Process Times
+		syncAPIProcessTime, err := stringToInt64(stats.Metrics.APIProcessTime.Sync.Current)
+		if err == nil {
```
Comment (maleck13) [inline pkg/mobile/metrics/fh-sync-server.go]: are we just ignoring errors here? Maybe what we could do is push any errors into a slice and after processing the stats, print any errors that occurred? Not a fan of ignoring the errors completely

File: pkg/mobile/metrics/fh-sync-server.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+	return ssMetrics, nil
+}
+
+func (ss *FhSyncServer) getStats(host string) (*statsResponse, error) {
+	statsURL := fmt.Sprintf("%s/sys/info/stats", host)
+	requester := ss.requestBuilder.Insecure(true).Build()
+	res, err := requester.Get(statsURL)
+	if err != nil {
+		return &statsResponse{}, err
```
Comment (maleck13) [inline pkg/mobile/metrics/fh-sync-server.go]: should probably return ```nil``` here

File: pkg/mobile/metrics/fh-sync-server_test.go [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+						},
+					}, nil
+				})
+				return client
+			},
+			Requester: func(t *testing.T) mobile.ExternalHTTPRequester {
+				return &mock.Requester{
+					Test: t,
+					Responder: func(host string, path string, method string, t *testing.T) (*http.Response, error) {
+						fmt.Print(path)
```
Comment (maleck13) [inline pkg/mobile/metrics/fh-sync-server_test.go]: can remove the println

File: pkg/mobile/metrics/fh-sync-server.go (hunk 3 above)
Comment (david-martin) [reply pkg/mobile/metrics/fh-sync-server.go]: From the api callers point of view, the metric can't be parsed, so best to leave it out.
Lets add logging on the server like you suggest for cases where the metric couldn't be parsed.

File: pkg/mobile/metrics/fh-sync-server.go (hunk 4 above)
Comment (david-martin) [reply pkg/mobile/metrics/fh-sync-server.go]: :+1:

File: pkg/mobile/metrics/fh-sync-server_test.go (hunk 5 above)
Comment (david-martin) [reply pkg/mobile/metrics/fh-sync-server_test.go]: :+1:

File: pkg/mobile/metrics/fh-sync-server.go (hunk 2 above)
Comment (david-martin) [reply pkg/mobile/metrics/fh-sync-server.go]: Yeah, lemme put something together that's  more robust at parsing the number

File: pkg/mobile/metrics/fh-sync-server.go (hunk 1 above)
Comment (david-martin) [reply pkg/mobile/metrics/fh-sync-server.go]: Agreed :+1:  It can be simplified a lot into just 4 structs that are re-used

**Comment by david-martin** [comment]
@maleck13 For your consideration based on the feedback 9ba1877

**Comment by maleck13** [comment]
updated the example

**Comment by david-martin** [comment]
@maleck13 Build passed with updates from your most recent suggestions. Going to merge

**Pull request merged**